### ISSUE TITLE
Saving previously seen nodes for later bootstrapping

### ISSUE THREAD

**Issue by dsvi** [ipfs/go-ipfs#3926]
#### Version information:
N/A

#### Type:
Enhancement
#### Severity:
Medium
#### Description:

Bootstrap nodes built into ipfs client don't seem to work for me at all. Had to find and add some manually. 
Generally having so centralized bootstrap system for otherwise decentralized network is its weakness:
https://github.com/ipfs/go-ipfs/issues/3908
Should ipfs automatically save all the nodes it seen and the time of last seen, to clean out the list later?
That should make the bootstrapping more reliable and truly decentralized.

**Comment by whyrusleeping** [comment]
Yes! definitely. The blocker here is making sure we store peerstore data to a persistent datastore (right now all peer info is kept in memory).

**Comment by djdv** [comment]
Related anecdote
[Perfect Dark](https://en.wikipedia.org/wiki/Perfect_Dark_(P2P)) relies on users to manually add nodes, the client retains peers like you had mentioned. It seems to work well for their network/peers. You add a peer to connect to, the client will exchange peer lists with them and add all of them to a ledger, on the next bootstrap the client uses that ledger, it attempts to establish a connection with everyone in it, after enough connection failures peers are removed from the ledger while successful ones remain, this keeps it from filling up with too many dead nodes.

Not sure what approach would be good for ipfs in terms of storing peers and managing stored peers for bootstrap, or what other systems have tried.

**Comment by whyrusleeping** [comment]
@djdv that sounds like a great approach. Especially tracking failures for removal.

**Comment by Cole128** [comment]
So did this happen? Thanks.

**Comment by makeworld-the-better-one** [comment]
@Kubuxu @whyrusleeping Hey it's me again. I'm just wondering if this was actually added in ipfs v0.4.12. It would be a useful feature and I can't find anything about it anywhere else. It seems @whyrusleeping added it, but maybe @Kubuxu removed it? This doesn't seem to be in the milestone anymore.

**Comment by makeworld-the-better-one** [comment]
@Kubuxu @whyrusleeping 
What's the state of this?

**Comment by Stebalien** [comment]
@bigs is working on storing the peerstore on disk (https://github.com/libp2p/go-libp2p-peerstore/issues/28). After that, it'll be a matter of remembering which peers tend to reliably be online.

**Comment by bigs** [comment]
yup. we'll also need to consider how we re-initialize our TTLs after rebooting, but this is coming down the pike.

**Comment by lidel** [comment]
Ideas being discussed in https://github.com/libp2p/go-libp2p-kad-dht/issues/254

**Comment by blurHY** [comment]
**GFW blocked all default bootstrap nodes**

> GFW is the national firewall of China

**Comment by Geo25rey** [comment]
@whyrusleeping Do you know the status of this issue?

**Comment by aschmahmann** [comment]
@Geo25rey there are a number of the issues (and some PRs) in the DHT around this that you can checkout.

The short version is that there's interest in doing this and some good proposals, but the plan is not to do it until we've landed work on performing smooth upgrades to the DHT protocol (e.g. https://github.com/libp2p/go-libp2p-kad-dht/issues/616).

**Comment by Geo25rey** [comment]
> we've landed higher priority work on performing smooth upgrades to the DHT protocol

@aschmahmann What do you mean by that?

**Comment by lidel** [comment]
Two years later, neither   https://github.com/libp2p/go-libp2p-kad-dht/issues/616 nor https://github.com/libp2p/go-libp2p-kad-dht/issues/254 happened.

Meanwhile, various countries and companies can cripple connectivity by blocking well-klnown list of hardcoded  bootstrappers.

I propose we do _something_ rather than keeping current broken state. 
Does not have to be perfect, MVP could live purely in go-ipfs code. 
When upstream libs support persistence and fancy logic, we can switch to them.

## Help wanted

If someone opens a PR that is preserving currently connected peers (`ipfs swarm connect`) across restarts, I'm happy to review it. 

MVP:
- persist the  list of connected peers every 15m / swarm peers is enumerated / on shutdown
- extend [bootstrap.go](https://github.com/ipfs/go-ipfs/blob/ef866a1400b3b2861e5e8b6cc9edc8633b890a0a/core/bootstrap/bootstrap.go#L172) logic to re-connect to persisted peers
  - use persisted peers as bootstrappers, but don't  add them with `peerstore.PermanentAddrTTL` - these should not be permanent bootstrappers
     - instead, use `peerstore.TempAddrTTL`  or `peerstore.ProviderAddrTTL` to ensure we keep them around for short period of time

**Comment by schomatis** [comment]
Per @lidel, taking this one since no one in the community has tackled it yet (if someone is still interested I can be the guide and reviewer but should say so now).

**Comment by schomatis** [comment]
Actively working on this. It's taking some time to do it right, will push something by EOW.

**Comment by schomatis** [comment]
Beautiful people of the ipfsphere, I have a very WIP PR with an initial implementation of this in https://github.com/ipfs/go-ipfs/pull/8856. Any feedback will be very useful to better understand the use cases we should support.
